File: concurrency/01115_print_foobar_alternately/spin.c
#include <stdatomic.h>
#include <inttypes.h>
#include <sched.h>

typedef struct {
    int n;
    atomic_int_least8_t state;
} FooBar;

// Function declarations. Do not change or remove this line
void printFoo();
void printBar();

FooBar* fooBarCreate(int n) {
    FooBar* obj = (FooBar*) malloc(sizeof(FooBar));
    atomic_init(&obj->state, 0);
    obj->n = n;
    return obj;
}

void foo(FooBar* obj) {
    for (int i = 0; i < obj->n; i++) {
        while (atomic_load_explicit(&obj->state, memory_order_acquire) != 0) {
            sched_yield();
        }
        // printFoo() outputs "foo". Do not change or remove this line.
        printFoo();
        atomic_store_explicit(&obj->state, 1, memory_order_release);
    }
}

void bar(FooBar* obj) {
    for (int i = 0; i < obj->n; i++) {
        while (atomic_load_explicit(&obj->state, memory_order_acquire) != 1) {
            sched_yield();
        }
        // printBar() outputs "bar". Do not change or remove this line.
        printBar();
        atomic_store_explicit(&obj->state, 0, memory_order_release);
    }
}

void fooBarFree(FooBar* obj) {
    free(obj);
}

File: concurrency/01115_print_foobar_alternately/spin.rs
use std::sync::atomic::{AtomicU8, Ordering};

struct FooBar {
    n: usize,
    state: AtomicU8,
}

impl FooBar {
    fn new(n: usize) -> Self {
        FooBar { n: n, state: AtomicU8::new(0) }
    }

    fn foo<F>(&self, print_foo: F)
    where
        F: Fn(),
    {
        for i in 0..self.n {
            while self.state.load(Ordering::Acquire) != 0 {
                std::hint::spin_loop();
            }
            // printFoo() outputs "foo". Do not change or remove this line.
            print_foo();
            self.state.store(1, Ordering::Release);
        }
    }

    fn bar<F>(&self, print_bar: F)
    where
        F: Fn(),
    {
        for i in 0..self.n {
            while self.state.load(Ordering::Acquire) != 1 {
                std::hint::spin_loop();
            }
            // printBar() outputs "bar". Do not change or remove this line.
            print_bar();
            self.state.store(0, Ordering::Release);
        }
    }
}
